sdks/python/apache_beam/transforms/online_clustering.py

```diff
@@ -0,0 +1,218 @@
+        clustering updates.
+
+          Example Usage::
+
+            pcoll | RunInference(
+                        XGBoostModelHandlerNumpy(
+                            model_class="XGBoost Model Class",
+                            model_state="my_model_state.json")))
+
+          Args:
+          n_clusters: number of clusters used by the algorithm
+          batch_size: size of the data batches
+          is_batched: boolean value that marks if the collection is already batched
+           and thus doesn't need to be batched by this transform
+          """
+    super().__init__()
+    self.n_clusters = n_clusters
+    self.batch_size = batch_size
+    self.is_batched = is_batched
+
+  def expand(self, pcoll):
+    pcoll = (
+        pcoll
+        | "Convert element to numpy arrays" >> beam.ParDo(TypeConversion()))
+
+    if not self.is_batched:
+      pcoll = (
+          pcoll
+          | "Create batches of elements" >> beam.BatchElements(
+              min_batch_size=self.n_clusters, max_batch_size=self.batch_size))
+
+    return (pcoll | "Add a key" >> beam.Map(lambda record: (1, record)))
+
+
+class OnlineClustering(ptransform.PTransform):
+  def __init__(
+      self,
+      clustering_algorithm,
+      n_clusters: int,
+      cluster_args: dict,
+      batch_size: int = 1024,
+      is_batched: bool = False):
+    """ Clustering transformation itself, it first preprocesses the data,
+        then it applies the clustering transformation step by step on each
+        of the batches.
+
+          Example Usage::
+
+            pcoll | OnlineClustering(
+                        clustering_algorithm=OnlineKMeansClustering
+                        batch_size=1024,
+                        n_clusters=6
+                        cluster_args={}))
+
+          Args:
+          clustering_algorithm: Clustering algorithm (DoFn)
+          n_clusters: Number of clusters
+          cluster_args: Arguments for the sklearn clustering algorithm (check sklearn documentation for more information)
+          batch_size: size of the data batches
+          is_batched: boolean value that marks if the collection is already batched
+           and thus doesn't need to be batched by this transform
+          """
+    super().__init__()
+    self.clustering_algorithm = clustering_algorithm
+    self.n_clusters = n_clusters
+    self.batch_size = batch_size
+    self.cluster_args = cluster_args
+    self.is_batched = is_batched
+
+  def expand(self, pcoll):
+    data = (
+        pcoll
+        | ClusteringPreprocessing(
+            n_clusters=self.n_clusters,
+            batch_size=self.batch_size,
+            is_batched=self.is_batched))
+
+    model = (
+        data
+        | 'Cluster' >> core.ParDo(
+            self.clustering_algorithm(
+                n_clusters=self.n_clusters, cluster_args=self.cluster_args))
+        | 'Select latest model state' >> core.CombineGlobally(
+            SelectLatestModelState()))
```

Comment:
I'm a little concerned with how this approach will interact with Beam's windowing. Specifically, in the scenario where you have multiple windows, since state is scoped to key/window we will discard all progress when we enter a new window (and we'll start a new clustering algorithm).

I think a better approach here might be to repeatedly emit the newly trained model and pick that value up in our side input - that way the data is always being labeled by the most recent version of the clustering algorithm.

This has some drawbacks - specifically, it will label data with potentially unknown versions of the model because side inputs aren't guaranteed to update before we process the data used to train the side input.

Reply:
I think regardless, we should probably move this to the examples directory for now (it can sit next to california_housing_clustering) since this story isn't fully fleshed out. I'd still request that we unlink the 2 pieces of the transform since I think its a better/more expressive pattern.

Putting it in the examples directory is probably for the best until we have a fuller suite of these algorithms anyways.